tests/runner-tests/file-data-sources/src/mapping.ts

```diff
@@ -4,130 +4,148 @@ import {
+    } else {
+      // This should never be logged if the entity was created in the offchain handler
+      // Such entities are not accessible in onchain handlers and will return null on load
+      log.info("Updating entity content", []);
+      entity.content = "updated content";
+      entity.save();
+    }
   }
 
-  let entity = new IpfsFile(dataSource.stringParam());
-  entity.content = data.toString();
-  entity.save();
-
-  // Test that an offchain data source can load its own entities
-  let loaded_entity = IpfsFile.load(dataSource.stringParam())!;
-  assert(loaded_entity.content == entity.content);
-}
+  if (event.params.testCommand == CREATE_FOO) {
+    let entity = new Foo(event.params.data);
+    entity.save();
+    let context = new DataSourceContext();
+    context.setString("command", CREATE_FOO);
+    dataSource.createWithContext("File", [event.params.data], context);
+  }
 
-export function handleFile1(data: Bytes): void {
-  let entity = new IpfsFile1(dataSource.stringParam());
-  entity.content = data.toString();
-  entity.save();
-}
+  if (event.params.testCommand == ACCESS_FILE_ENTITY_THROUGH_DERIVED_FIELD) {
+    let entity = Foo.load(event.params.data);
+    if (entity == null) {
+      log.info("Entity not found", []);
+    } else {
+      log.info("Accessing file entity through derived field", []);
+      let fileEntity = entity.ipfs.load();
 
-// Used to test spawning a file data source from another file data source handler.
-// This function spawns a file data source that will be handled by `handleSpawnedTest`.
-export function spawnTestHandler(data: Bytes): void {
-  let context = new DataSourceContext();
-  context.setString("file", "fromSpawnTestHandler");
-  let command = dataSource.context().getString("command");
-  if (command == "spawnOffChainHandlerTest") {
-    dataSource.createWithContext("OffChainDataSource", [KNOWN_CID], context);
-  } else if (command == "spawnOnChainHandlerTest") {
-    dataSource.createWithContext("OnChainDataSource", [KNOWN_CID], context);
+      assert(fileEntity.length == 0, "Expected exactly one file entity");
+    }
   }
 }
 
-// This is the handler for the data source spawned by `spawnOffChainHandlerTest`.
-export function handleSpawnedTest(data: Bytes): void {
-  let entity = new SpawnTestEntity(dataSource.stringParam());
-  let context = dataSource.context().getString("file");
-  entity.content = data.toString();
-  entity.context = context;
-  entity.save();
+export function handleFile(data: Bytes): void {
+  log.info("handleFile {}", [dataSource.stringParam()]);
+  let context = dataSource.context();
+
+  if (context.isSet("command")) {
+    let contextCommand = context.getString("command");
+
+    if (contextCommand == SPAWN_FDS_FROM_OFFCHAIN_HANDLER) {
+      let hash = context.getString("hash");
+      log.info("Creating file data source from handleFile: {}", [hash]);
+      dataSource.createWithContext("File", [hash], new DataSourceContext());
+    }
```

Comment:
Not sure what's idiomatic for TypeScript bu wouldn't `else if` look more like a switch statement?

Reply:
Agree, it should have been an else if.

We cannot use switch cases though since switch-case in assemblyscript does not work with string
https://github.com/AssemblyScript/assemblyscript/issues/1643